File: 066.Plus One.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int head = digits.size() - 1;
        if(digits.size() == 0)
            return digits;
            
        int flag = (digits[head] + 1) / 10;
        digits[head] = (digits[head] + 1) % 10;
        for (int i = head-1; i >= 0; i--) {
            int sum = flag + digits[i];
            digits[i] = sum % 10;
            flag = sum / 10;
        }
        
        if (flag != 0) {
            // vector<int>::iterator iterator = digits.begin();
            digits.insert(digits.begin(), flag);
        }
        
        return digits;
    }
};

// ============ improve ================

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();
    	for (int i = n - 1; i >= 0; --i)
    	{
    		if (digits[i] == 9)
    		{
    			digits[i] = 0;
    		}
    		else
    		{
    			digits[i]++;
    			return digits;
    		}
    	}
    	
    	digits[0] = 1;
    	digits.push_back(0);
        
        return digits;
    }
};

File: 066.Plus One.java
/*
    Given a non-negative integer represented as a non-empty array of digits, plus one to the integer.

    You may assume the integer do not contain any leading zero, except the number 0 itself.

    The digits are stored such that the most significant digit is at the head of the list.
     */

    public int[] plusOne(int[] digits) {
        int sum;
        sum = digits[digits.length-1] + 1;
        digits[digits.length-1] = sum % 10;
        int carry = sum / 10;
        for (int i=digits.length-2; i>=0; i--) {
            sum = digits[i] + carry;
            digits[i] = sum % 10;
            carry = sum / 10;
        }

        if (carry > 0) {
            int size = digits.length + 1;
            int[] results = new int[size];
            results[0] = carry;
            for (int i=0; i<digits.length; i++) {
                results[i+1] = digits[i];
            }

            return results;
        }

        return digits;
    }
